src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -147,656 +104,784 @@ private void FinalizeDispose()
-            /// watches must be explicitly added for those subdirectories.
-            /// </summary>
-            private readonly bool _includeSubdirectories;
-            /// <summary>
-            /// Token to monitor for cancellation requests, upon which processing is stopped and all
-            /// state is cleaned up.
-            /// </summary>
-            private readonly CancellationToken _cancellationToken;
-            /// <summary>
-            /// Mapping from watch descriptor (as returned by inotify_add_watch) to state for
-            /// the associated directory being watched.  Events from inotify include only relative
-            /// names, so the watch descriptor in an event must be used to look up the associated
-            /// directory path in order to convert the relative filename into a full path.
-            /// </summary>
-            private readonly Dictionary<int, WatchedDirectory> _wdToPathMap = new Dictionary<int, WatchedDirectory>();
-            /// <summary>
-            /// Maximum length of a name returned from inotify event.
-            /// </summary>
-            private const int NAME_MAX = 255; // from limits.h
+            private WatchedDirectory[] _dirBuffer = new WatchedDirectory[4];
 
-            /// <summary>Initializes the instance with all state necessary to operate a watch.</summary>
-            internal RunningInstance(
-                FileSystemWatcher watcher, SafeFileHandle inotifyHandle, string directoryPath,
-                bool includeSubdirectories, NotifyFilters notifyFilters, CancellationToken cancellationToken)
+            public INotify(object watcherLock)
             {
-                Debug.Assert(watcher != null);
-                Debug.Assert(inotifyHandle != null && !inotifyHandle.IsInvalid && !inotifyHandle.IsClosed);
-                Debug.Assert(directoryPath != null);
-
-                _weakWatcher = new WeakReference<FileSystemWatcher>(watcher);
-                _inotifyHandle = inotifyHandle;
-                _directoryPath = directoryPath;
-                _buffer = watcher.AllocateBuffer();
-                Debug.Assert(_buffer != null && _buffer.Length > (c_INotifyEventSize + NAME_MAX + 1));
-                _includeSubdirectories = includeSubdirectories;
-                _notifyFilters = notifyFilters;
-                _watchFilters = TranslateFilters(notifyFilters);
-                _cancellationToken = cancellationToken;
-
-                // Add a watch for this starting directory.  We keep track of the watch descriptor => directory information
-                // mapping in a dictionary; this is needed in order to be able to determine the containing directory
-                // for all notifications so that we can reconstruct the full path.
-                AddDirectoryWatchUnlocked(null, directoryPath);
+                _watchersLock = watcherLock;
+
```

Comment:
This lock is always `INotify.s_watchersLock`. Since it's static and always the same value, perhaps it would be simpler to just use `INotify.s_watchersLock` everywhere? It would be easier to find all the usages and it would be clear it's the exact same lock guard